v-next/hardhat-ledger/src/internal/hook-handlers/network.ts

```diff
@@ -0,0 +1,117 @@
+import type { HookContext, NetworkHooks } from "hardhat/types/hooks";
+import type { ChainType, NetworkConnection } from "hardhat/types/network";
+import type { JsonRpcRequest, JsonRpcResponse } from "hardhat/types/providers";
+
+import { assertHardhatInvariant } from "@nomicfoundation/hardhat-errors";
+import { AsyncMutex } from "@nomicfoundation/hardhat-utils/synchronization";
+
+import { LedgerHandler } from "../handler.js";
+import { isFailedJsonRpcResponse, isJsonRpcResponse } from "../rpc-helpers.js";
+
+export default async (): Promise<Partial<NetworkHooks>> => {
+  // This map is essential for managing multiple network connections in Hardhat V3.
+  // Since Hardhat V3 supports multiple connections, we use this map to track each one
+  // and associate it with the corresponding handlers array.
+  // When a connection is closed, its associated handler is removed from the map.
+  // See the "closeConnection" function at the end of the file for more details.
+  const ledgerHandlerPerConnection: WeakMap<
+    NetworkConnection<ChainType | string>,
+    LedgerHandler
+  > = new Map();
```

Comment:
```suggestion
  const ledgerHandlerPerConnection: Map<
    NetworkConnection<ChainType | string>,
    LedgerHandler
  > = new Map();
```
I think we are just using it as a map, as we need to do explicit resource cleanup. Or do you think both type and instance should be `WeakMap`?

Reply:
We use the same logic in the `handlers`: https://github.com/NomicFoundation/hardhat/blob/d22f9dd3543a9762b8cb438fe436795267f7551a/v-next/hardhat/src/internal/builtin-plugins/network-manager/hook-handlers/network.ts#L23

I think that having a `WeakMap` is a safe net to clean up resources if for any reason `closeConnection` is not properly called. See old comment: https://github.com/NomicFoundation/hardhat/pull/5994#discussion_r1871977938

Reply:
Done here: https://github.com/NomicFoundation/hardhat/pull/6983/commits/5f59a872da958a0f838cd46aca4a03f867b50aa3